Generalize KNN Image Classifier to KNN Classifier, taking an arbitrary embedding as input.

nsthorat opened the pull request:
This PR generalizes the KNN Image Classifier to be a generic KNN Classifier for any embedding.

This PR also:
- Removes the dependency on MobileNet. It is now the user's responsibility to pass a MobileNet embedding to the model.

cc @HalfdanJ -- feel free to review, for some reason I can't add you to reviewers.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tensorflow/tfjs-models/36)
<!-- Reviewable:end -->

HalfdanJ commented:
Review status: 0 of 1 LGTMs obtained

---

*[knn-classifier/src/index.ts, line 140 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNKK3wDDlAV54caKAl:-LFNKK3wDDlAV54caKAm:bvdbfbt) ([raw file](https://github.com/tensorflow/tfjs-models/blob/b73956fbd5f2da246619ba7f27ed5a4cd4f4c26d/knn-classifier/src/index.ts#L140)):*
> ```XML
>    * values for all possible classes.
>    */
>   async predictClass(example: Tensor, k = 3):
> ```

It's a little confusing to me that the variable here is called example. Should it be image or logits instead? Also, the comments still have a lot of references to image, but now that the classifier isnt limited to images, should that be abstracted so its more clear it can be used for more? 

---

*[knn-classifier/demo/camera.js, line 146 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNJr1Y-_58fV0d3PSY:-LFNJr1Y-_58fV0d3PSZ:bwtenv6) ([raw file](https://github.com/tensorflow/tfjs-models/blob/b73956fbd5f2da246619ba7f27ed5a4cd4f4c26d/knn-classifier/demo/camera.js#L146)):*
> ```JavaScript
> 
>   // If any examples have been added, run predict
>   const exampleCount = classifier.getNumClasses();
> ```

Looking at the implementation of getNumClasses, I can see this will work, but it seems like a confusing variable name now. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/36)*
<!-- Sent from Reviewable.io -->

nsthorat commented:
Review status: 0 of 1 LGTMs obtained

---

*[knn-classifier/src/index.ts, line 140 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNKK3wDDlAV54caKAl:-LFNMUxO6sAy1muv7dym:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/b73956fbd5f2da246619ba7f27ed5a4cd4f4c26d/knn-classifier/src/index.ts#L140)):*
<details><summary><i>Previously, HalfdanJ (Jonas Jongejan) wrote…</i></summary><blockquote>

It's a little confusing to me that the variable here is called example. Should it be image or logits instead? Also, the comments still have a lot of references to image, but now that the classifier isnt limited to images, should that be abstracted so its more clear it can be used for more? 
</blockquote></details>

Done.

---

*[knn-classifier/demo/camera.js, line 146 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNJr1Y-_58fV0d3PSY:-LFNMUSg1hKsYQlvG8VS:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/b73956fbd5f2da246619ba7f27ed5a4cd4f4c26d/knn-classifier/demo/camera.js#L146)):*
<details><summary><i>Previously, HalfdanJ (Jonas Jongejan) wrote…</i></summary><blockquote>

Looking at the implementation of getNumClasses, I can see this will work, but it seems like a confusing variable name now. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/36)*
<!-- Sent from Reviewable.io -->

HalfdanJ commented:
Review status: 0 of 1 LGTMs obtained

---

*[a discussion](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNVvJm9lxshEaAejQ-:-LFNVvJm9lxshEaAejQ0:b-p5c59r) (no related file):*
Sorry, couple extra comments (implementing the code into the work im doing at the moment) 

---

*[knn-classifier/demo/camera.js, line 131 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNVYzt12TztjPm1IaC:-LFNVYzt12TztjPm1IaD:br64v5h) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/demo/camera.js#L131)):*
> ```JavaScript
>  * Animation function called on each frame, running prediction
>  */
> async function animate() {
> ```

Im getting a `[Violation] 'requestAnimationFrame' handler took 138ms` warning in the console in chrome. Any tricks to workaround that?

---

*[knn-classifier/demo/camera.js, line 137 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNVey51pfJysTJaory:-LFNVey51pfJysTJaorz:b-48954u) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/demo/camera.js#L137)):*
> ```JavaScript
>   const image = tf.fromPixels(video);
>   let logits;
>   const infer = () => mobilenet.infer(image, 'conv_preds');
> ```

Can you add a comment to why `conv_preds`? 

---

*[knn-classifier/demo/README.md, line 60 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNVEfM5X10-Jtfqr63:-LFNVEfM5X10-Jtfqr64:b6074wc) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/demo/README.md#L60)):*
> ```Markdown
> Link the local knn-classifier to the demos:
> ```sh
> yalc link \@tensorflow-models/knn-classifier
> ```

You changed the demo to now use a local reference to the module, so this isn't necessary anymore. But not sure if it was intentional to change. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/36)*
<!-- Sent from Reviewable.io -->

caisq commented:
Review status: 0 of 1 LGTMs obtained

---

*[knn-classifier/README.md, line 13 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNT90m8Bk4wJ9EpHTY:-LFNT90m8Bk4wJ9EpHTZ:bai442j) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L13)):*
> ```
> Usage
> ```

nit: "Usage example", to be precise.

---

*[knn-classifier/README.md, line 26 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNTIZgEBeKyQ2gAsLi:-LFNTIZgEBeKyQ2gAsLj:b76r4j4) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L26)):*
> ```
> Add embeddings
> ```

IIUC, the following adds only one embedding, instead of mulitple? Should the example show how to add multiple embeddings, to be more general?

---

*[knn-classifier/README.md, line 26 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNU59y0kVQyoqrr8Ly:-LFNU59y0kVQyoqrr8Lz:b-j7banc) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L26)):*
> ```
>  embedding
> ```

Another comment: is "embedding" standard terminology for talking about kNN classifiers? I find it potentially confusing because people may confuse it with word embedding. (Yes, they have similar meanings. But it may be still worthwhile defining / explaining the word somewhere.)

---

*[knn-classifier/README.md, line 46 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNUQnd8NoCu2ZBTdsF:-LFNUQnd8NoCu2ZBTdsG:b-bpmij4) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L46)):*
> ```
> knnClassifier.create()
> ```

Why create(), instead of 1) a constructor call or 2) `knnClassifier()`? Constructors are frowned upon in JS APIs. But 2 seems to be more concise and friendly?

---

*[knn-classifier/README.md, line 56 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNVT74DwKZYWs3g-z0:-LFNVT74DwKZYWs3g-z1:b-wyai5i) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L56)):*
> ```
> classIndex: number
> ```

Does this have to be contiguous numbers? For example, what happens if the user provides 0, 2, 4, 6, 8, without providing 1, 3, 5, 7?

Does it make sense for this to be `number|string`? May be the API will be easier to use that way?

---

*[knn-classifier/README.md, line 70 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNUsHRDjNEdLX_ikDh:-LFNUsHRDjNEdLX_ikDi:blm1f0t) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L70)):*
> ```
> k = 3
> ```

What happens if the number of examples available is < 3?

---

*[knn-classifier/README.md, line 89 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNVBSf-UUPkDvA2iKd:-LFNVBSf-UUPkDvA2iKe:b3yvcre) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L89)):*
> ```
> clearClass(
> ```

Does it make sense to provide a method to clear all examples from all classes?

---

*[knn-classifier/README.md, line 128 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNVKXY5YsVGpdEW-Fe:-LFNVKXY5YsVGpdEW-Ff:b-vnamu) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L128)):*
> ```
> classifier.dispose()
> ```

This frees WebGL memory? If so, may be worth mentioning that.

---

*[knn-classifier/src/index.ts, line 1 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNW2aBBPLey9tLu4Z9:-LFNW2aBBPLey9tLu4ZA:bj6ab0j) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/src/index.ts#L1)):*
> ```
> /**
> ```

Does it make sense / would it be worthwhile adding some unit tests for this module?

---

*[knn-classifier/src/index.ts, line 41 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNWOe28SQKB6LQ34d1:-LFNWOe28SQKB6LQ34d2:b-oixppz) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/src/index.ts#L41)):*
> ```
> example.shape
> ```

Is there any requirement on the dimensionality of the shape? For example, thinking about edge cases, would a Scalar shape work? 

---

*[knn-classifier/src/index.ts, line 53 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNWFw-8C10nQG7wPAh:-LFNWFw-8C10nQG7wPAi:b-o2y00o) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/src/index.ts#L53)):*
> ```
> this.clearTrainDatasetMatrix();
> ```

Does this mean this has a O(n^2) complexity, n being the number examples ever added? If so, is there a way to optimize it?

---

*[knn-classifier/src/index.ts, line 82 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNWoMvDr03XRFQAI0h:-LFNWoMvDr03XRFQAI0i:b-l3mk12) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/src/index.ts#L82)):*
> ```
> returns the K-nearest neighbors
> ```

Below it says the method returns cosine distances. Here it says it returns neighbors. So which one is right?

---

*[knn-classifier/src/index.ts, line 128 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNYBu97mmvyWqnwsF3:-LFNYBu97mmvyWqnwsF4:bk9rmni) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/src/index.ts#L128)):*
> ```
>  const knn = this.knn(input).asType('float32');
> ```

This function is not using a tidy() inside. Is it because tidy() doesn't work with async? Double check everything is cleaned up here?

---

*[knn-classifier/src/index.ts, line 176 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNX6VTAYs_mm48ZlX-:-LFNX6VTAYs_mm48ZlX0:b-ywci6f) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/src/index.ts#L176)):*
> ```
>  kVal
> ```

Throw an error if kVal is <= 0?

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/36)*
<!-- Sent from Reviewable.io -->

nsthorat commented:
Review status: 0 of 1 LGTMs obtained

---

*[knn-classifier/README.md, line 13 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNT90m8Bk4wJ9EpHTY:-LFNYqM2Cm3GeECXMtTu:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L13)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> Usage
> ```

nit: "Usage example", to be precise.
</blockquote></details>

Done.

---

*[knn-classifier/README.md, line 26 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNTIZgEBeKyQ2gAsLi:-LFNZ1SUC7WnOnCz3CvU:b-pa0eb3) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L26)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> Add embeddings
> ```

IIUC, the following adds only one embedding, instead of mulitple? Should the example show how to add multiple embeddings, to be more general?
</blockquote></details>

I added a comment to make it clear that this is meant to be repeated (I didn't want to bloat this since it's clear, I think).

---

*[knn-classifier/README.md, line 26 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNU59y0kVQyoqrr8Ly:-LFNZJYYEu_YCeIJI8bJ:bnxj5cn) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L26)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
>  embedding
> ```

Another comment: is "embedding" standard terminology for talking about kNN classifiers? I find it potentially confusing because people may confuse it with word embedding. (Yes, they have similar meanings. But it may be still worthwhile defining / explaining the word somewhere.)
</blockquote></details>

Now using "activation" instead of embedding.

---

*[knn-classifier/README.md, line 46 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNUQnd8NoCu2ZBTdsF:-LFNZNAb5WD1ZrF_WAwo:bsk2kqs) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L46)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> knnClassifier.create()
> ```

Why create(), instead of 1) a constructor call or 2) `knnClassifier()`? Constructors are frowned upon in JS APIs. But 2 seems to be more concise and friendly?
</blockquote></details>

With the latter it becomes:

`knnClassifier.knnClassifier();`

I wanted to avoid this, thoughts?

---

*[knn-classifier/README.md, line 56 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNVT74DwKZYWs3g-z0:-LFNZTM411lA50bjtRnc:b-s9ei9c) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L56)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> classIndex: number
> ```

Does this have to be contiguous numbers? For example, what happens if the user provides 0, 2, 4, 6, 8, without providing 1, 3, 5, 7?

Does it make sense for this to be `number|string`? May be the API will be easier to use that way?
</blockquote></details>

Nope, you can provide whatever non-contiguous numbers. It actually complicates things a bit to add strings to the API, and I figure that it's pretty simple for users to have their own class mapping.

---

*[knn-classifier/README.md, line 70 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNUsHRDjNEdLX_ikDh:-LFNZsU-1uR2qkFFR5sD:bn5ofqe) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L70)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> k = 3
> ```

What happens if the number of examples available is < 3?
</blockquote></details>

Then k becomes the number of examples, I added some text to specify that.

---

*[knn-classifier/README.md, line 89 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNVBSf-UUPkDvA2iKd:-LFNbd_SCVdz_zXtKBWy:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L89)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> clearClass(
> ```

Does it make sense to provide a method to clear all examples from all classes?
</blockquote></details>

Done.

---

*[knn-classifier/README.md, line 128 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNVKXY5YsVGpdEW-Fe:-LFN_FS9DFhgvaeo3cpu:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L128)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> classifier.dispose()
> ```

This frees WebGL memory? If so, may be worth mentioning that.
</blockquote></details>

Done.

---

*[knn-classifier/src/index.ts, line 1 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNW2aBBPLey9tLu4Z9:-LFN_Yfx9S1rxwIf5ego:b-ylxl93) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/src/index.ts#L1)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> /**
> ```

Does it make sense / would it be worthwhile adding some unit tests for this module?
</blockquote></details>

It definitely makes sense -- I'll set this all up in a follow up since it requires a non-trivial amount of wiring for this repo.

---

*[knn-classifier/src/index.ts, line 41 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNWOe28SQKB6LQ34d1:-LFN_c60AA2Rc7QDMq0A:b-yl2klz) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/src/index.ts#L41)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> example.shape
> ```

Is there any requirement on the dimensionality of the shape? For example, thinking about edge cases, would a Scalar shape work? 
</blockquote></details>

Yes, a scalar shape should work, since we call flatten() which returns a Tensor1D. There are no requirements.

---

*[knn-classifier/src/index.ts, line 53 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNWFw-8C10nQG7wPAh:-LFN_mM-EMHMub8n0O6y:bev3fbi) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/src/index.ts#L53)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> this.clearTrainDatasetMatrix();
> ```

Does this mean this has a O(n^2) complexity, n being the number examples ever added? If so, is there a way to optimize it?
</blockquote></details>

Adding an example is not O(n^2). clearTrainDatasetMatrix is so that we lazily construct the concatenated matrix, this is an invalidation. Adding an example adds a single row to the matrix (hard to measure complexity of this since it's parallelized). Then when the user makes a prediction, we create a new train dataset matrix which is the concatenation of all the individual class matrices. This only happens the first time after an example to a class gets added.

---

*[knn-classifier/src/index.ts, line 82 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNWoMvDr03XRFQAI0h:-LFNaTgi5Tfol2ePOI0i:bv5gzfj) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/src/index.ts#L82)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> returns the K-nearest neighbors
> ```

Below it says the method returns cosine distances. Here it says it returns neighbors. So which one is right?
</blockquote></details>

Done. Nice catch. It's just similarities.

---

*[knn-classifier/src/index.ts, line 128 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNYBu97mmvyWqnwsF3:-LFNaijm-q-ENFJwQ4tU:b2pu9wp) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/src/index.ts#L128)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
>  const knn = this.knn(input).asType('float32');
> ```

This function is not using a tidy() inside. Is it because tidy() doesn't work with async? Double check everything is cleaned up here?
</blockquote></details>

Done. Good catch, was leaking 1 tensor.

---

*[knn-classifier/src/index.ts, line 176 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNX6VTAYs_mm48ZlX-:-LFNaqQx4LQHog94xVQe:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/src/index.ts#L176)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
>  kVal
> ```

Throw an error if kVal is <= 0?
</blockquote></details>

Done.

---

*[knn-classifier/demo/camera.js, line 131 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNVYzt12TztjPm1IaC:-LFN_FwrCbGrw2RfGKAs:b-mt0key) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/demo/camera.js#L131)):*
<details><summary><i>Previously, HalfdanJ (Jonas Jongejan) wrote…</i></summary><blockquote>

Im getting a `[Violation] 'requestAnimationFrame' handler took 138ms` warning in the console in chrome. Any tricks to workaround that?
</blockquote></details>

Unfortunately not really, we'd have to make the model smaller :/

---

*[knn-classifier/demo/camera.js, line 137 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNVey51pfJysTJaory:-LFN_OEdE_lQosrnWBrB:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/demo/camera.js#L137)):*
<details><summary><i>Previously, HalfdanJ (Jonas Jongejan) wrote…</i></summary><blockquote>

Can you add a comment to why `conv_preds`? 
</blockquote></details>

Done.

---

*[knn-classifier/demo/README.md, line 60 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNVEfM5X10-Jtfqr63:-LFN_Vo56kuySqyc2Bgh:b3dwea2) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/demo/README.md#L60)):*
<details><summary><i>Previously, HalfdanJ (Jonas Jongejan) wrote…</i></summary><blockquote>

You changed the demo to now use a local reference to the module, so this isn't necessary anymore. But not sure if it was intentional to change. 
</blockquote></details>

Done. It was intentional.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/36)*
<!-- Sent from Reviewable.io -->

nsthorat commented:
Review status: :shipit: complete! 1 of 1 LGTMs obtained

---

*[knn-classifier/README.md, line 9 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNSn3n-KscIPSwzm77:-LFNivmf87sATcahXpG1:b-k7744o) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L9)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

i think 'from' is missing here. But i also think there may be a different way of expressing this last part. How about something like

"...but rather a utility for constructing a classifier that can be used on the output (or internal activations) of another model, or any other categorical data that you have."
</blockquote></details>

Updated the text (this isn't really for categorical data).

---

*[knn-classifier/README.md, line 75 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNVc0d0kCUQHPaTffZ:-LFNjCLO9QhYFFDDByLL:b-me3gg0) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L75)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

Small comment. I wonder if calling this example might throw people off wrt to the examples passed to addExample?
</blockquote></details>

Done. Good catch, I actually renamed it in the API but not in the README.

---

*[knn-classifier/README.md, line 77 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNV5oL3rw9kbdQQgsb:-LFNjPTs6dVuuuoTjAIR:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L77)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

How about" 

"The K value to use in K-nearest neighbors. The algorithm will first find the K nearest examples from those it was previously shown, and then choose the class that appears the most as the final prediction for the input example. Defaults to 3"
</blockquote></details>

Done.

---

*[knn-classifier/README.md, line 81 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNWIKu-NUQoz6xT0dk:-LFNjVeEDNoYKoY9099B:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L81)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

Might want to clarify that this is for all the classes.
</blockquote></details>

Done.

---

*[knn-classifier/README.md, line 106 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNTrKf01qSHYPq-jWe:-LFNjgzC0DMhqQH8J5Zw:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L106)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

I wonder if this could just be called `getClassifierDataset` and the setter called `setClassifierDataset`
</blockquote></details>

Done.

---

*[knn-classifier/README.md, line 126 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNTUXs5IYfvWlCWia4:-LFNjidp68_zB2LOrSSh:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L126)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

Do you think it would be usefule to talk more about when you might want to use this? Even something like, 'if you no longer need to use the classifier but your application will continue to run use this method'?
</blockquote></details>

Done.

---

*[knn-classifier/src/index.ts, line 29 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFN_6i25MIGvuN5eEJY:-LFNkGo309IPuiPMJCWR:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/src/index.ts#L29)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

might be nice to add a comment(s) here, particularly to distinguish `trainDataMatrix` and `classDatasetMatrices`.  You can also describe the on demand creation of this during the knn step. 
</blockquote></details>

Done.

---

*[knn-classifier/src/index.ts, line 145 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNaT5n7bxmzaEAUce-:-LFNkMjo63mU-jiSvoP_:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/src/index.ts#L145)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

do you need this setting to null, and the setting to 0 beloew since you call `delete` on the property?
</blockquote></details>

Done.

---

*[knn-classifier/src/index.ts, line 188 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNcEk-61DTI3_R3X7t:-LFNkd4P72L7H50DB8sx:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/src/index.ts#L188)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

Since this is not used in this loop but in the next one. I wonder if something like `const topKCountsForClasses = Array(Object.keys(this.classDatasetMatrices).length).fill(0);` might allow you to move this line closer to the next loop where it is used.
</blockquote></details>

Done.

---

*[knn-classifier/src/index.ts, line 190 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNatoV8MpD-rGslrm1:-LFNkPAO3c9KXOeMqqza:bwzn99q) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/src/index.ts#L190)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

I don't understand why the `+i > 0` is here and the `+i - 1` below.
</blockquote></details>

Since classDatasetMatices is a map of number => value, i is actually a string.

---

*[knn-classifier/src/index.ts, line 197 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNcw8yAaw_nLbOkgeR:-LFNlNP19gBESuxi3NPw:bqsyvuy) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/src/index.ts#L197)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

might another name for `classForEntry` be `classIndex`. Reading the first name made me think 'what are the entries?' and had me looking for something i had missed. 

It might be nice to describe what topKIndices contains, maybe in a comment at the top of this method. Else its a bit hard to trace through what this double loop is doing. The intuition of what it looks to be doing seems apparent, but its somewhat hard to follow when just reading this function definition
</blockquote></details>

Done.  Add a jsdoc for topKIndices.

---

*[mobilenet/package.json, line 34 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNXwjH63-1AQ_X7sMi:-LFNkmOtBjBNjUcit619:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/mobilenet/package.json#L34)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

Not necessarily related to your change, but it doesn't look like this should work. Might be old?
</blockquote></details>

Done.

---

*[knn-classifier/demo/camera.js, line 114 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNUOuyB54RpvFgRvWm:-LFNk-7nDSSDzyK61OwF:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/demo/camera.js#L114)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

This function name seems confusing as it is primarily loading mobileNet, it also happens to create and return a classifer, but that isn't really loaded. Suggest refactoring to separate the two (or to remove the function)
</blockquote></details>

Done.

---

*[knn-classifier/demo/camera.js, line 170 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNX-yv6pYMuO-wLZb5:-LFNk0Aq7LQGNw3yz6QH:b-xpyfs6) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/demo/camera.js#L170)):*
<details><summary><i>Previously, tafsiri (Yannick Assogba) wrote…</i></summary><blockquote>

consider `if (logits)`
</blockquote></details>

It's better to keep explicit != null checks in general since "logits" could be falsy

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/36)*
<!-- Sent from Reviewable.io -->

nsthorat commented:
Thanks for all the excellent comments!

HalfdanJ commented:
Review status: :shipit: complete! 1 of 1 LGTMs obtained

---

*[knn-classifier/demo/README.md, line 60 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNVEfM5X10-Jtfqr63:-LFNpVib-63nEsU-gGEI:bvc9fad) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/demo/README.md#L60)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

Done. It was intentional.
</blockquote></details>

It makes sense it has been changed to local reference, but then you dont need to use yalc? 

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/36)*
<!-- Sent from Reviewable.io -->

nsthorat commented:
Review status: :shipit: complete! 1 of 1 LGTMs obtained

---

*[knn-classifier/demo/README.md, line 60 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNVEfM5X10-Jtfqr63:-LFO66fQCCCz99Xze2M7:b-fqwmts) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/demo/README.md#L60)):*
<details><summary><i>Previously, HalfdanJ (Jonas Jongejan) wrote…</i></summary><blockquote>

It makes sense it has been changed to local reference, but then you dont need to use yalc? 
</blockquote></details>

Removed all references to yalc

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/36)*
<!-- Sent from Reviewable.io -->

caisq commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Review status: :shipit: complete! 2 of 1 LGTMs obtained

---

*[knn-classifier/README.md, line 26 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNTIZgEBeKyQ2gAsLi:-LFOEYtN6k1jR0nqdFHn:b-1p7dqp) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L26)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

I added a comment to make it clear that this is meant to be repeated (I didn't want to bloat this since it's clear, I think).
</blockquote></details>

Why not show code that adds two examples (of two categories). It'll probably be more concise than the words you added here.

---

*[knn-classifier/README.md, line 46 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNUQnd8NoCu2ZBTdsF:-LFOEigLE2B_-fHMQmjO:b-7hs1vr) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L46)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

With the latter it becomes:

`knnClassifier.knnClassifier();`

I wanted to avoid this, thoughts?
</blockquote></details>

Good point. I guess `knnClassifier.create()` is fine.

---

*[knn-classifier/README.md, line 56 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNVT74DwKZYWs3g-z0:-LFOEqyG8V2L_0t4chhb:b-aq78yc) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L56)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

Nope, you can provide whatever non-contiguous numbers. It actually complicates things a bit to add strings to the API, and I figure that it's pretty simple for users to have their own class mapping.
</blockquote></details>

OK. If so, does the `addExample` method check that class is really a number and throw an error if it isn't?

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-:-LFOFBYm7B270T6y0hiW:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

nsthorat commented:
Review status: :shipit: complete! 2 of 1 LGTMs obtained

---

*[knn-classifier/README.md, line 26 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNTIZgEBeKyQ2gAsLi:-LFOJHTVA_xrUZKh2ej8:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L26)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

Why not show code that adds two examples (of two categories). It'll probably be more concise than the words you added here.
</blockquote></details>

Done.

---

*[knn-classifier/README.md, line 46 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNUQnd8NoCu2ZBTdsF:-LFOJIP7AamqbjPKCy4Y:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L46)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

Good point. I guess `knnClassifier.create()` is fine.
</blockquote></details>

Done.

---

*[knn-classifier/README.md, line 56 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/36#-LFNVT74DwKZYWs3g-z0:-LFOJMIh-oagFpFfpUx3:b39gvlg) ([raw file](https://github.com/tensorflow/tfjs-models/blob/02ae6a4b1d0470ff4af21e9513683e4add6b65d9/knn-classifier/README.md#L56)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

OK. If so, does the `addExample` method check that class is really a number and throw an error if it isn't?
</blockquote></details>

Yep it does!

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/36)*
<!-- Sent from Reviewable.io -->
